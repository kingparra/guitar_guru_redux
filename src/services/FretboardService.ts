import { TUNING, NUM_FRETS, ALL_NOTES, NOTE_MAP } from '../constants';
import type { DiagramNote, FingeringMap, PathDiagramNote } from '../types';
import { HandPositionModel } from '../models/HandPositionModel';

type ScaleNote = { noteName: string; degree: string };

export class FretboardService {

    public static generateNotesOnFretboard(scaleNotes: ScaleNote[]): DiagramNote[] {
        const notes: DiagramNote[] = [];
        if (!scaleNotes?.length) return notes;
        const scaleNoteMap = new Map(scaleNotes.map(n => [n.noteName, n.degree]));
        for (let stringIndex = 0; stringIndex < TUNING.length; stringIndex++) {
            const openStringNoteIndex = NOTE_MAP[TUNING[stringIndex].toUpperCase()];
            if (openStringNoteIndex === undefined) continue;
            for (let fret = 0; fret <= NUM_FRETS; fret++) {
                const currentNoteName = ALL_NOTES[(openStringNoteIndex + fret) % ALL_NOTES.length];
                if (scaleNoteMap.has(currentNoteName)) {
                    notes.push({ string: stringIndex, fret, noteName: currentNoteName, degree: scaleNoteMap.get(currentNoteName)! });
                }
            }
        }
        return notes;
    }

    public static generateFingeringPositions(notesOnFretboard: DiagramNote[]): FingeringMap[] {
        if (notesOnFretboard.length < 7) return [];
        const positions: FingeringMap[] = [];
        const processedWindows = new Set<string>();

        for (let fret = 0; fret <= 15; fret++) {
            const hasAnchorNote = notesOnFretboard.some(note => (note.string === 6 || note.string === 5) && note.fret === fret);
            if (hasAnchorNote) {
                const windowA_Start = fret;
                const windowB_Start = Math.max(0, fret - 1);
                const countNotes = (start: number, end: number) => notesOnFretboard.filter(n => typeof n.fret === 'number' && n.fret >= start && n.fret <= end).length;
                const noteCountA = countNotes(windowA_Start, windowA_Start + 4);
                const noteCountB = countNotes(windowB_Start, windowB_Start + 4);

                const baseFret = (noteCountB > noteCountA && fret > 0) ? windowB_Start : windowA_Start;
                const maxFret = baseFret + 4;
                const windowKey = `${baseFret}-${maxFret}`;
                if (processedWindows.has(windowKey)) continue;

                const notesInPosition = notesOnFretboard.filter(note => typeof note.fret === 'number' && note.fret >= baseFret && note.fret <= maxFret);
                const positionFingering: FingeringMap = notesInPosition.map(note => {
                    const fret = note.fret as number;
                    const finger = (fret === 0 && baseFret === 0) ? '0' : (fret - baseFret + 1).toString();
                    return { key: `${note.string}_${fret}`, finger };
                });

                const stringsCovered = new Set(notesInPosition.map(n => n.string)).size;
                if (stringsCovered >= 5 && positionFingering.length >= 10) {
                    positions.push(positionFingering);
                    processedWindows.add(windowKey);
                }
            }
        }
        positions.sort((a, b) => {
            const minFretA = Math.min(...a.map(item => parseInt(item.key.split('_')[1], 10)));
            const minFretB = Math.min(...b.map(item => parseInt(item.key.split('_')[1], 10)));
            return minFretA - minFretB;
        });
        return positions.slice(0, 7);
    }

    public static generateDiagonalRun(notesOnFretboard: DiagramNote[]): PathDiagramNote[] {
        if (notesOnFretboard.length === 0) return [];
        const handModel = new HandPositionModel(notesOnFretboard);
        const lowestRoot = notesOnFretboard
            .filter(n => n.degree === 'R')
            .sort((a, b) => b.string - a.string || (typeof a.fret === 'number' && typeof b.fret === 'number' ? a.fret - b.fret : 0))[0];
        return lowestRoot ? handModel.generateFullPath(lowestRoot) : [];
    }

    public static formatPathAsTab(path: PathDiagramNote[]): { columns: {string: number, fret: string}[][] } {
        if (!path || path.length === 0) return { columns: [] };
        const tab: { columns: {string: number, fret: string}[][] } = { columns: [] };
        const NOTES_PER_BAR = 8;

        for (let i = 0; i < path.length; i++) {
            const note = path[i];
            if (typeof note.string !== 'number' || note.string < 0 || note.string >= TUNING.length) {
                console.warn('Skipping note with invalid string index:', note);
                continue;
            }
            const prevNote = i > 0 ? path[i - 1] : null;
            const tabColumn = Array.from({ length: 7 }, (_, s) => ({ string: s, fret: '-' }));
            let fretValue = note.fret.toString();
            if (note.shiftType === 'slide' && prevNote && prevNote.string === note.string) {
                 fretValue = `/${note.fret}`;
            }
            tabColumn[note.string].fret = fretValue;
            tab.columns.push(tabColumn);

            if ((i + 1) % NOTES_PER_BAR === 0 && i < path.length - 1) {
                tab.columns.push(Array.from({ length: 7 }, (_, s) => ({ string: s, fret: '|' })));
            }
        }
        return tab;
    }
}
