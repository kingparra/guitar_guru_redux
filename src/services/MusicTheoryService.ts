import { SCALE_FORMULAS, ALL_NOTES, NOTE_MAP } from '../constants';
import type { Result } from '../types';

type ScaleNote = { noteName: string; degree: string };

const PROGRESSION_FORMULAS: Record<string, { name: string; analysis: string; degrees: string[]; }[]> = {
    'Major': [
        { name: 'Classic Pop/Rock', analysis: 'I - V - vi - IV', degrees: ['I', 'V', 'vi', 'IV'] },
        { name: 'Folk & Blues', analysis: 'I - IV - V', degrees: ['I', 'IV', 'V'] },
    ],
    'Natural Minor': [
        { name: 'Standard Minor', analysis: 'i - VI - III - VII', degrees: ['i', 'VI', 'III', 'VII'] },
        { name: 'Andalusian Cadence', analysis: 'i - VII - VI - V', degrees: ['i', 'VII', 'VI', 'V'] },
    ],
    'Harmonic Minor': [
        { name: 'Dramatic Minor', analysis: 'i - iv - V - i', degrees: ['i', 'iv', 'V', 'i'] },
        { name: 'Neoclassical Turnaround', analysis: 'i - VI - vii° - i', degrees: ['i', 'VI', 'vii°', 'i'] },
    ],
};

export class MusicTheoryService {
    public static generateScaleNotes(rootNote: string, scaleName: string): Result<ScaleNote[], Error> {
        const formula = SCALE_FORMULAS[scaleName];
        if (!formula) return { type: 'failure', error: new Error(`Scale formula for "${scaleName}" not found.`) };
        
        const scaleNotes: ScaleNote[] = [{ noteName: rootNote, degree: 'R' }];
        let currentNoteIndex = NOTE_MAP[rootNote];
        for (const [interval, degree] of formula) {
            currentNoteIndex = (currentNoteIndex + interval) % ALL_NOTES.length;
            scaleNotes.push({ noteName: ALL_NOTES[currentNoteIndex], degree });
        }
        return { type: 'success', value: scaleNotes };
    }

    public static getCharacteristicDegrees(scaleNotes: ScaleNote[]): string[] {
        const degrees = new Set(scaleNotes.map(n => n.degree));
        const characteristicDegrees = ['b2', '#4', 'b6'].filter(d => degrees.has(d));
        if (degrees.has('7') && degrees.has('b6')) characteristicDegrees.push('7');
        if (degrees.has('6') && degrees.has('b7')) characteristicDegrees.push('6');
        return characteristicDegrees;
    }
    
    public static generateDegreeTableMarkdown(scaleNotes: ScaleNote[]): string {
        const headers = '| Degree | Note |';
        const separator = '|---|---|';
        const rows = scaleNotes.map(n => `| ${n.degree} | ${n.noteName} |`).join('\n');
        return `${headers}\n${separator}\n${rows}`;
    }

    public static getProgressionFormulas(scaleName: string) {
        return PROGRESSION_FORMULAS[scaleName] || PROGRESSION_FORMULAS['Natural Minor']; // Default
    }
}