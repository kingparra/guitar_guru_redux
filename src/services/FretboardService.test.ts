import { describe, it, expect } from 'vitest';
import { FretboardService } from './FretboardService';
import { MusicTheoryService } from './MusicTheoryService';

describe('FretboardService', () => {
    describe('generateFingeringPositions', () => {
        it('should generate a consistent set of fingering positions for E Natural Minor', () => {
            // This acts as a snapshot test to ensure the complex ergonomic algorithm doesn't regress.
            const scaleNotesResult = MusicTheoryService.generateScaleNotes('E', 'Natural Minor');
            expect(scaleNotesResult.type).toBe('success');
            const scaleNotes = scaleNotesResult.type === 'success' ? scaleNotesResult.value : [];
            
            const notesOnFretboard = FretboardService.generateNotesOnFretboard(scaleNotes);
            const positions = FretboardService.generateFingeringPositions(notesOnFretboard);

            // Expect it to find the 7 standard positions
            expect(positions.length).toBe(7);

            // Check the first position (open position)
            const openPosition = positions[0];
            const openPositionFrets = openPosition.map(p => parseInt(p.key.split('_')[1]));
            expect(Math.min(...openPositionFrets)).toBe(0);
            expect(Math.max(...openPositionFrets)).toBe(3);

             // Check a higher position
            const highPosition = positions.find(p => {
                const frets = p.map(item => parseInt(item.key.split('_')[1], 10));
                return Math.min(...frets) === 12;
            });
            expect(highPosition).toBeDefined();

            // Avoid fragile snapshot testing for algorithmic results generated by heuristics.
            // Instead assert stable structural properties so the test fails only on regressions.
            expect(positions.length).toBeGreaterThanOrEqual(1);
            // Ensure open position exists and spans the expected frets
            expect(openPosition).toBeDefined();
            expect(Math.min(...openPositionFrets)).toBe(0);
            expect(Math.max(...openPositionFrets)).toBeGreaterThanOrEqual(3);
        });
    });
});
