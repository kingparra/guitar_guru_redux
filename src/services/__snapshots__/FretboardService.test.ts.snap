// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FretboardService > generateFingeringPositions > should generate a consistent set of fingering positions for E Natural Minor 1`] = `
[
  [
    {
      "finger": "0",
      "key": "0_0",
    },
    {
      "finger": "3",
      "key": "0_2",
    },
    {
      "finger": "4",
      "key": "0_3",
    },
    {
      "finger": "0",
      "key": "1_0",
    },
    {
      "finger": "2",
      "key": "1_1",
    },
    {
      "finger": "4",
      "key": "1_3",
    },
    {
      "finger": "0",
      "key": "2_0",
    },
    {
      "finger": "3",
      "key": "2_2",
    },
    {
      "finger": "5",
      "key": "2_4",
    },
    {
      "finger": "0",
      "key": "3_0",
    },
    {
      "finger": "3",
      "key": "3_2",
    },
    {
      "finger": "5",
      "key": "3_4",
    },
    {
      "finger": "0",
      "key": "4_0",
    },
    {
      "finger": "3",
      "key": "4_2",
    },
    {
      "finger": "4",
      "key": "4_3",
    },
    {
      "finger": "0",
      "key": "5_0",
    },
    {
      "finger": "3",
      "key": "5_2",
    },
    {
      "finger": "4",
      "key": "5_3",
    },
    {
      "finger": "0",
      "key": "6_0",
    },
    {
      "finger": "2",
      "key": "6_1",
    },
    {
      "finger": "4",
      "key": "6_3",
    },
  ],
  [
    {
      "finger": "2",
      "key": "0_2",
    },
    {
      "finger": "3",
      "key": "0_3",
    },
    {
      "finger": "5",
      "key": "0_5",
    },
    {
      "finger": "1",
      "key": "1_1",
    },
    {
      "finger": "3",
      "key": "1_3",
    },
    {
      "finger": "5",
      "key": "1_5",
    },
    {
      "finger": "2",
      "key": "2_2",
    },
    {
      "finger": "4",
      "key": "2_4",
    },
    {
      "finger": "5",
      "key": "2_5",
    },
    {
      "finger": "2",
      "key": "3_2",
    },
    {
      "finger": "4",
      "key": "3_4",
    },
    {
      "finger": "5",
      "key": "3_5",
    },
    {
      "finger": "2",
      "key": "4_2",
    },
    {
      "finger": "3",
      "key": "4_3",
    },
    {
      "finger": "5",
      "key": "4_5",
    },
    {
      "finger": "2",
      "key": "5_2",
    },
    {
      "finger": "3",
      "key": "5_3",
    },
    {
      "finger": "5",
      "key": "5_5",
    },
    {
      "finger": "1",
      "key": "6_1",
    },
    {
      "finger": "3",
      "key": "6_3",
    },
    {
      "finger": "5",
      "key": "6_5",
    },
  ],
  [
    {
      "finger": "1",
      "key": "0_3",
    },
    {
      "finger": "3",
      "key": "0_5",
    },
    {
      "finger": "5",
      "key": "0_7",
    },
    {
      "finger": "1",
      "key": "1_3",
    },
    {
      "finger": "3",
      "key": "1_5",
    },
    {
      "finger": "5",
      "key": "1_7",
    },
    {
      "finger": "2",
      "key": "2_4",
    },
    {
      "finger": "3",
      "key": "2_5",
    },
    {
      "finger": "5",
      "key": "2_7",
    },
    {
      "finger": "2",
      "key": "3_4",
    },
    {
      "finger": "3",
      "key": "3_5",
    },
    {
      "finger": "5",
      "key": "3_7",
    },
    {
      "finger": "1",
      "key": "4_3",
    },
    {
      "finger": "3",
      "key": "4_5",
    },
    {
      "finger": "5",
      "key": "4_7",
    },
    {
      "finger": "1",
      "key": "5_3",
    },
    {
      "finger": "3",
      "key": "5_5",
    },
    {
      "finger": "5",
      "key": "5_7",
    },
    {
      "finger": "1",
      "key": "6_3",
    },
    {
      "finger": "3",
      "key": "6_5",
    },
    {
      "finger": "5",
      "key": "6_7",
    },
  ],
  [
    {
      "finger": "1",
      "key": "0_5",
    },
    {
      "finger": "3",
      "key": "0_7",
    },
    {
      "finger": "4",
      "key": "0_8",
    },
    {
      "finger": "1",
      "key": "1_5",
    },
    {
      "finger": "3",
      "key": "1_7",
    },
    {
      "finger": "4",
      "key": "1_8",
    },
    {
      "finger": "1",
      "key": "2_5",
    },
    {
      "finger": "3",
      "key": "2_7",
    },
    {
      "finger": "5",
      "key": "2_9",
    },
    {
      "finger": "1",
      "key": "3_5",
    },
    {
      "finger": "3",
      "key": "3_7",
    },
    {
      "finger": "5",
      "key": "3_9",
    },
    {
      "finger": "1",
      "key": "4_5",
    },
    {
      "finger": "3",
      "key": "4_7",
    },
    {
      "finger": "5",
      "key": "4_9",
    },
    {
      "finger": "1",
      "key": "5_5",
    },
    {
      "finger": "3",
      "key": "5_7",
    },
    {
      "finger": "4",
      "key": "5_8",
    },
    {
      "finger": "1",
      "key": "6_5",
    },
    {
      "finger": "3",
      "key": "6_7",
    },
    {
      "finger": "4",
      "key": "6_8",
    },
  ],
  [
    {
      "finger": "1",
      "key": "0_7",
    },
    {
      "finger": "2",
      "key": "0_8",
    },
    {
      "finger": "4",
      "key": "0_10",
    },
    {
      "finger": "1",
      "key": "1_7",
    },
    {
      "finger": "2",
      "key": "1_8",
    },
    {
      "finger": "4",
      "key": "1_10",
    },
    {
      "finger": "1",
      "key": "2_7",
    },
    {
      "finger": "3",
      "key": "2_9",
    },
    {
      "finger": "5",
      "key": "2_11",
    },
    {
      "finger": "1",
      "key": "3_7",
    },
    {
      "finger": "3",
      "key": "3_9",
    },
    {
      "finger": "4",
      "key": "3_10",
    },
    {
      "finger": "1",
      "key": "4_7",
    },
    {
      "finger": "3",
      "key": "4_9",
    },
    {
      "finger": "4",
      "key": "4_10",
    },
    {
      "finger": "1",
      "key": "5_7",
    },
    {
      "finger": "2",
      "key": "5_8",
    },
    {
      "finger": "4",
      "key": "5_10",
    },
    {
      "finger": "1",
      "key": "6_7",
    },
    {
      "finger": "2",
      "key": "6_8",
    },
    {
      "finger": "4",
      "key": "6_10",
    },
  ],
  [
    {
      "finger": "1",
      "key": "0_8",
    },
    {
      "finger": "3",
      "key": "0_10",
    },
    {
      "finger": "5",
      "key": "0_12",
    },
    {
      "finger": "1",
      "key": "1_8",
    },
    {
      "finger": "3",
      "key": "1_10",
    },
    {
      "finger": "5",
      "key": "1_12",
    },
    {
      "finger": "2",
      "key": "2_9",
    },
    {
      "finger": "4",
      "key": "2_11",
    },
    {
      "finger": "5",
      "key": "2_12",
    },
    {
      "finger": "2",
      "key": "3_9",
    },
    {
      "finger": "3",
      "key": "3_10",
    },
    {
      "finger": "5",
      "key": "3_12",
    },
    {
      "finger": "2",
      "key": "4_9",
    },
    {
      "finger": "3",
      "key": "4_10",
    },
    {
      "finger": "5",
      "key": "4_12",
    },
    {
      "finger": "1",
      "key": "5_8",
    },
    {
      "finger": "3",
      "key": "5_10",
    },
    {
      "finger": "5",
      "key": "5_12",
    },
    {
      "finger": "1",
      "key": "6_8",
    },
    {
      "finger": "3",
      "key": "6_10",
    },
    {
      "finger": "5",
      "key": "6_12",
    },
  ],
  [
    {
      "finger": "1",
      "key": "0_10",
    },
    {
      "finger": "3",
      "key": "0_12",
    },
    {
      "finger": "5",
      "key": "0_14",
    },
    {
      "finger": "1",
      "key": "1_10",
    },
    {
      "finger": "3",
      "key": "1_12",
    },
    {
      "finger": "4",
      "key": "1_13",
    },
    {
      "finger": "2",
      "key": "2_11",
    },
    {
      "finger": "3",
      "key": "2_12",
    },
    {
      "finger": "5",
      "key": "2_14",
    },
    {
      "finger": "1",
      "key": "3_10",
    },
    {
      "finger": "3",
      "key": "3_12",
    },
    {
      "finger": "5",
      "key": "3_14",
    },
    {
      "finger": "1",
      "key": "4_10",
    },
    {
      "finger": "3",
      "key": "4_12",
    },
    {
      "finger": "5",
      "key": "4_14",
    },
    {
      "finger": "1",
      "key": "5_10",
    },
    {
      "finger": "3",
      "key": "5_12",
    },
    {
      "finger": "5",
      "key": "5_14",
    },
    {
      "finger": "1",
      "key": "6_10",
    },
    {
      "finger": "3",
      "key": "6_12",
    },
    {
      "finger": "4",
      "key": "6_13",
    },
  ],
]
`;
